django/db/models/query.py

```diff
@@ -608,7 +608,12 @@ def update_or_create(self, defaults=None, **kwargs):
                 return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
-            obj.save(using=self.db)
+            # `update_fields` does not support non-concrete fields yet (#31382)
+            fnames = map(lambda f: f.name, self.model._meta.concrete_fields)
+            if set(defaults).issubset(fnames):
+                obj.save(using=self.db, update_fields=defaults.keys())
```

Comment:
`update_fields` can be any iterable, `iter({}) -> dict_keyiterator`

```suggestion
                obj.save(using=self.db, update_fields=defaults)
```